src/test/kotlin/platformapi/ChannelDetailsTests.kt

```diff
@@ -0,0 +1,209 @@
+package platformapi
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import dev.saibotma.persistence.postgres.jooq.tables.Channel.Companion.CHANNEL
+import dev.saibotma.persistence.postgres.jooq.tables.pojos.Channel
+import error.ApiError
+import error.ApiException
+import error.duplicate
+import io.kotest.matchers.collections.shouldBeIn
+import io.kotest.matchers.collections.shouldHaveSize
+import io.kotest.matchers.shouldBe
+import io.kotest.matchers.shouldNotBe
+import io.ktor.http.*
+import io.ktor.server.testing.*
+import di.setupKodein
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.toList
+import kotlinx.coroutines.runBlocking
+import module
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import org.kodein.di.*
+import org.kodein.di.ktor.closestDI
+import org.testcontainers.containers.PostgreSQLContainer
+import persistence.postgres.ChatServerPostgres
+import persistence.postgres.PostgresConfig
+import platformapi.models.ChannelWrite
+import platformapi.models.toChannelWrite
+import testutil.DatabaseTest
+import testutil.mockedChannelWrite
+
+class ChannelDetailsTests : DatabaseTest() {
+    @Nested
+    inner class UpsertChannelTests {
+        @Test
+        fun `creates a channel when no one exists yet`() {
+            serverTest {
+                val channel = upsertChannel(mockedChannelWrite())
+
+                with(getChannels().map { it.toChannelWrite() }.toList()) {
+                    shouldHaveSize(1)
+                    first() shouldBe channel
+                }
+            }
+        }
+
+        @Test
+        fun `updates a channel when it already exists`() {
+            serverTest {
+                val channel = upsertChannel(mockedChannelWrite())
+                val updatedChannel = upsertChannel(channel.copy(name = "updated name", isManaged = !channel.isManaged))
+
+                with(getChannels().map { it.toChannelWrite() }.toList()) {
+                    shouldHaveSize(1)
+                    first() shouldBe updatedChannel
+                }
+            }
+        }
+    }
+}
+
+fun DI.MainBuilder.setupTestDependencies() {
+    val postgresContainer = PostgreSQLContainer<Nothing>()
```

Comment:
What is the generic type here? :O (normally)

Reply:
It's a recursive generic type. 
`public class PostgreSQLContainer<SELF extends PostgreSQLContainer<SELF>> extends JdbcDatabaseContainer<SELF>`. Do you have any idea on how to make this work?